sherpa/stats/__init__.py

```diff
@@ -64,10 +68,173 @@ def __repr__(self):
+            raise StatErr('mismatch',
+                          'number of data sets', ndata,
+                          'model expressions', nmdl)
+
+    def _validate_inputs(self, data, model):
+        """Ensure that the inputs are correct for the statistic.
+
+        The default behavior is to check that the data contains
+        at least one bin and that the number of datasets matches
+        the number of models. It also converts single values to
+        simultaneous objects, if necessary.
+
+        Parameters
+        ----------
+        data : a Data or DataSimulFit instance
+            The data set, or sets, to use.
+        model : a Model or SimulFitModel instance
+            The model expressions for each data set. It must match
+            the data parameter (the models are in the same order
+            as the data objects).
+
+        Returns
+        -------
+        data, model : DataSimulFit instance, SimulFitModel instance
+            If the input was a SimulFit object then this is just
+            the input value.
+        """
+
+        if self._calc is None:
+            # This is a programmer error rather than a user error,
+            # so use NotImplementedError rather than
+            # StatErr('nostat', self.name, '_calc')
+            #
+            raise NotImplementedError("_calc method has not been set")
+
+        data, model = self._bundle_inputs(data, model)
+        self._check_has_bins(data)
+        self._check_sizes_match(data, model)
+        return data, model
+
+    # TODO:
+    #  - should this accept sherpa.data.Data input instead of
+    #    "raw" data (i.e. to match calc_stat)
+    #  - should this be moved out of the base Stat class since
+    #    isn't relevant for likelihood statistics?
+    #
     def calc_staterror(self, data):
+        """Return the statistic error values for the data.
+
+        Parameters
+        ----------
+        data : scalar or 1D array of numbers
+            The data values.
+
+        Returns
+        -------
+        staterror : scalar or array of numbers
+            The errors for the input data values (matches the data
+            argument).
+
+        """
         raise NotImplementedError
 
-    def calc_stat(self, data, model, staterror, *args, **kwargs):
+    # TODO: add *args, **kwargs?
+    def calc_stat(self, data, model):
+        """Return the statistic value for the data and model.
+
+        Parameters
+        ----------
+        data : a Data or DataSimulFit instance
+            The data set, or sets, to use.
+        model : a Model or SimulFitModel instance
+            The model expression, or expressions. If a SimulFitModel
+            is given then it must match the number of data sets in the
+            data parameter.
+
+        Returns
+        -------
+        statval, fvec : number, array of numbers
+            The statistic value and the per-bin "statistic" value.
+
+        """
+
         raise NotImplementedError
```

Comment:
Should we consider using proper abstract base classes? I guess this is a question for another day.